File: ts/src/compose.ts
import { additive } from "./additive.js";
import {
  type AdditiveRequest,
  type MinimumRequest,
  parseRequest,
  type ScoreRequest,
  type ScoreResult,
  type WeightedMeanRequest,
} from "./contracts.js";
import { minimum } from "./minimum.js";
import { weightedMean } from "./weightedMean.js";

export function compose(input: WeightedMeanRequest): ScoreResult;
export function compose(input: AdditiveRequest): ScoreResult;
export function compose(input: MinimumRequest): ScoreResult;
export function compose(input: ScoreRequest): ScoreResult;
export function compose(input: ScoreRequest): ScoreResult {
  const request = parseRequest(input);
  if (request.method === "weighted_mean") return weightedMean(request);
  if (request.method === "additive") return additive(request);
  return minimum(request);
}

File: src/assay/compose.py
"""One explicit dispatcher for Assay's closed composition methods."""

from __future__ import annotations

from typing import overload

from assay.additive import additive
from assay.contracts import AdditiveRequest, MinimumRequest, ScoreResult, WeightedMeanRequest
from assay.minimum import minimum
from assay.weighted_mean import weighted_mean


@overload
def compose(request: WeightedMeanRequest) -> ScoreResult: ...


@overload
def compose(request: AdditiveRequest) -> ScoreResult: ...


@overload
def compose(request: MinimumRequest) -> ScoreResult: ...


def compose(request: WeightedMeanRequest | AdditiveRequest | MinimumRequest) -> ScoreResult:
    """Dispatch one validated request without accepting executable formulas."""
    if isinstance(request, WeightedMeanRequest):
        return weighted_mean(request)
    if isinstance(request, AdditiveRequest):
        return additive(request)
    return minimum(request)
